File: Grind75_1/week7/WordSearch.py
"""
https://leetcode.com/problems/word-search/
"""

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        r = len(board)

        if (r == 0):
            return False

        c = len(board[0])
        lenW = len(word)

        visited = [[False] * c for _ in range(r)]

        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

        def DFS(row:int, col:int, wordIdx: int) -> bool:

            if (row < 0 or row >= r or col < 0 or col >= c
                or visited[row][col] == True
                or word[wordIdx] != board[row][col]):
                return False
            elif (lenW - 1 == wordIdx and word[wordIdx] == board[row][col]):
                return True

            found = False
            # mark visited so that the tile cannot be used again in the solution
            visited[row][col] = True

            for ro, co in directions:
                # check in all directions
                found = DFS(row + ro, col + co, wordIdx + 1)
                
                if (found):
                    return found
            
            # after tile evaluation, mark it false so it can be used in another solution
            visited[row][col] = False
            return found


        found = False
        for x in range(r):
            for y in range(c):
                if (board[x][y] == word[0]):
                    # only start DFS if first letter matches
                    found = DFS(x, y, 0)
                    if (found):
                        return found

        return found

File: Grind75/Graph/word-search.js
// https://leetcode.com/problems/word-search/description/

/**
main
    create visited of same dimensions of board and fill with false

    iterate rows
        iterate cols
            if board[r][c] === first character of the word
                if (rec(...) === true) // search if the word can be found at this starting point
                    return true

    return false

rec(...)
* board
* {int} r
* {int} c
* rows
* cols
* word
* i // current index of word
* visited
* directions

    base case 1:
    if (i >= word.length)
        // no more chars to check, meaning that all previous have matched
        return true

    base case 2:
    if r or c out of bounds or visited[r][c] === true or board[r][c] not word[i]
        return false

    visited[r][c] = true    // so not re-used on this path
    i += 1  // to next char in word

    for (directions)
        if (rec(...) === true)
            // if word found, propagate out
            return true

    visited[r][c] = false   // since path using this cell finished, set to false so another path can use it
    return false

- Time: O(r * c)
- Space: O(r * c)
 */

/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function(board, word) {
    const rows = board.length
    const cols = board[0].length
    const visited = Array.from(new Array(rows), (e) => {return new Array(cols).fill(false)})
    const directions = [[-1, 0], [0, 1], [1, 0], [0, -1]]

    for (let r = 0; r < rows; r ++) {
        for (let c = 0; c < cols; c ++) {
            if (word[0] === board[r][c]) {
                if (rec(board, word, r, c, 0, rows, cols, visited, directions) === true) {
                    return true
                }
            }
        }
    }

    return false
};

const rec = (board, word, r, c, i, rows, cols, visited, directions) => {
    if (i >= word.length) {
        return true
    }
    if (r < 0 || r >= rows || c < 0 || c >= cols || visited[r][c] === true || board[r][c] !== word[i]) {
        return false
    }

    visited[r][c] = true
    i += 1

    for (let [dr, dc] of directions) {
        if (rec(board, word, r + dr, c + dc, i, rows, cols, visited, directions) === true) {
            return true
        }
    }

    visited[r][c] = false
    return false
}